base/iterators.jl

```diff
@@ -957,4 +958,97 @@ function next(itr::PartitionIterator, state)
     return resize!(v, i), state
 end
 
+"""
+    Stateful(itr)
+
+There are several different ways to think about this iterator wrapper:
+    1. It provides a mutable wrapper around an iterator and
+       its iteration state.
+    2. It turns an iterator-like abstraction into a Channel-like
+       abstraction.
+    3. It's an iterator that mutates to become its own rest iterator
+       whenever an item is produced.
+
+`Stateful` provides the regular iterator interface. Like other mutable iterators
+(e.g. Channel), if iteration is stopped early (e.g. by a `break` in a for loop),
+iteration can be resumed from the same spot by continuing to iterate over the
+same iterator object (in contrast, an immutable iterator would restart from the
+beginning).
+
+# Example:
+```jldoctest
+julia> a = Iterators.Stateful("abcdef");
+
+julia> isempty(a)
+false
+
+julia> take!(a)
+'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
+
+julia> collect(Iterators.take(a, 3))
+3-element Array{Char,1}:
+ 'b'
+ 'c'
+ 'd'
+
+julia> collect(a)
+2-element Array{Char,1}:
+ 'e'
+ 'f'
+```
+
+```jldoctest
+julia> a = Iterators.Stateful([1,1,1,2,3,4]);
+
+# Skip any leading ones
+julia> for x in a; x == 1 || break; end
+
+# Sum the remaining elements
+julia> sum(a)
+7
+````
+"""
+struct Stateful{VS,T}
+    itr::T
+    # A bit awkward right now, but adapted to the new iteration protocol
+    nextvalstate::Ref{Union{VS, Nothing}}
+    taken::Ref{Int}
```

Comment:
![taken](http://s2.quickmeme.com/img/4a/4a3e4f9e2c8194f18e6a0ed55ba9b0ccee9469efdb0a6fa08e51a282c0e0e200.jpg)

Reply:
This field counts how many "Taken" movies have been released at the time of execution. It is currently set to 3.

Reply:
I don't expect that using a ref-cell will necessarily be a performance win. In most cases, we'll probably end up stripping it away, but if we don't, it's an extra memory indirection.

Reply:
A syntax I had proposed once upon a time was the equivalent of this:
```
mutable struct Stateful{VS,T}
    const itr::T
    state::Union{VS, Nothing}
    taken::Int
end
```
which would disallow assignment to the `itr` field and let the compiler assume that it can never change without forcing the other two fields to be ref cells. Seems like an occasionally useful feature. We could plan on doing that and just leave this as mutable now and document that the `itr` field should never be changed.